### PR TITLE
Tgui port

### PR DESCRIPTION
**Pull request by Loganbacca**
Big thanks to @neersighted for his amazing rewrite of NanoUI

Todo:
 - [x] Copy all tgui code
 - [x] Fix initial errors (compile ready)
 - [x] Port SMES, APC, and Solar Control tgui
 - [x] Remove NanoUI

I think that's enough for this PR.
Tgui is working and nanoui has been removed.
It will still need to be cleaned up (I created a bit of a mess just to get it to compile), but I'll come back and do that. I want to take a break from porting and ui's and work on something else for a bit *cough* lighting *cough*.

### CODE REVIEW

**Comment by PsiOmegaDelta** [comment]
/json_encode() might be a /tg/ proc but tgui may have been made 510 reliant.

**Comment by neersighted** [comment]
We have our own json writer. All 510 procs are stubbed so our code runs on both 509 and 510. See 510.dm for our implementation in DM of 510's json_encode.

**Comment by neersighted** [comment]
So, really excited to see tgui coming to a non-direct downstream! If you have any questions, bugs, requests etc let me know. I'd love to help with the porting.

The only UI that should require a major refactor is cargo as I rewrote most of the ordering code. Otherwise, converting NanoUI procs is fairly straightforward. Also its quite novel to have tgui and NanoUI -- unexpected result of the rename.

File: code/modules/power/smes.dm [hunk 1]
```diff
@@ -343,10 +370,70 @@
...
 		// auto update every Master Controller tick
 		ui.set_auto_update(1)
-
+*/
 /obj/machinery/power/smes/proc/Percentage()
 	return round(100.0*charge/capacity, 0.1)
 
+/obj/machinery/power/smes/ui_act(action, params)
+	if(..())
+		return
```
Comment (PsiOmegaDelta) [inline code/modules/power/smes.dm]: Should probably ```return 1``` for consistency even if you may never have SMES subtypes.

File: code/modules/tgui/states/default.dm [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	. = shared_ui_interaction(src_object)
+	if(. < UI_INTERACTIVE)
+		return
+
+	// The AI can interact with anything it can see nearby, or with cameras.
+	if((get_dist(src, src_object) <= client.view) || cameranet.checkTurfVis(get_turf(src_object)))
+		return UI_INTERACTIVE
+	return UI_CLOSE
+
+/mob/living/simple_animal/drone/default_can_use_topic(src_object)
```
Comment (PsiOmegaDelta) [inline code/modules/tgui/states/default.dm]: Bay-derivate drones are currently children of /silicon/robot.

**Comment by Loganbacca** [comment]
I'll return to this shortly after my 510 port is done and merged.

Thanks @neersighted I'll let you know if I have any questions.

@PsiOmegaDelta thanks, I'll get those sorted out.

**Comment by Zuhayr** [comment]
@neersighted tgui looks really really nice! I am technically inept so I can't really offer any feedback other than that.

I think offhand the only nanoUIs that we will need to convert are APCs and SMES units, and the solar control console, since almost all other station infrastructure has either been removed or is irrelevant, and the replacement machinery uses a very basic HTML interface at the moment.

**Comment by neersighted** [comment]
Well all three of those are tgui on /tg/, so it should be easy... Glad you like it!

**Comment by Loganbacca** [comment]
SMES, APC, and solar control are all ported to tgui.

The NanoUI JSON writer has been removed, so we are now running purely on 510's built in json procs.

I've started removing NanoUI, but it's embeded pretty deep in the code base, so this may take a while.

**Comment by Loganbacca** [comment]
Not compiling yet, but down to about 13 errors. Slowly getting there.

Lots of stuff I've just commented out until it is compiling. In time they will be either replaced with tgui or removed entirely.

**Comment by Loganbacca** [comment]
Ready for review.

**Comment by Zuhayr** [comment]
As discussed on IRC: move fast and break stuff.

**Pull request merged**